fullnode/fullnode_lazy.go

```diff
@@ -0,0 +1,235 @@
+	blockClient, err := newBlockClient(config.RpcURL)
+	if err != nil {
+		return nil, fmt.Errorf("NewSdk: error creating new Shannon block client at URL %s: %w", config.RpcURL, err)
+	}
+
+	config.GRPCConfig = config.GRPCConfig.hydrateDefaults()
+
+	sessionClient, err := newSessionClient(config.GRPCConfig)
+	if err != nil {
+		return nil, fmt.Errorf("NewSdk: error creating new Shannon session client using URL %s: %w", config.GRPCConfig.HostPort, err)
+	}
+
+	appClient, err := newAppClient(config.GRPCConfig)
+	if err != nil {
+		return nil, fmt.Errorf("NewSdk: error creating new GRPC connection at url %s: %w", config.GRPCConfig.HostPort, err)
+	}
+
+	accountClient, err := newAccClient(config.GRPCConfig)
+	if err != nil {
+		return nil, fmt.Errorf("NewSdk: error creating new account client using url %s: %w", config.GRPCConfig.HostPort, err)
+	}
+
+	fullNode := &LazyFullNode{
+		sessionClient: sessionClient,
+		appClient:     appClient,
+		blockClient:   blockClient,
+		accountClient: accountClient,
+	}
+
+	return fullNode, nil
+}
+
+// GetApp:
+// - Returns the onchain application matching the supplied application address.
+// - Required to fulfill the FullNode interface.
+func (lfn *LazyFullNode) GetApp(ctx context.Context, appAddr string) (*apptypes.Application, error) {
+	app, err := lfn.appClient.GetApplication(ctx, appAddr)
+	return &app, err
+}
+
+// GetSession:
+// - Uses the Shannon SDK to fetch a session for the (serviceID, appAddr) combination.
+// - Required to fulfill the FullNode interface.
+func (lfn *LazyFullNode) GetSession(
+	ctx context.Context,
+	serviceID sdk.ServiceID,
+	appAddr string,
+) (sessiontypes.Session, error) {
+	session, err := lfn.sessionClient.GetSession(
+		ctx,
+		appAddr,
+		string(serviceID),
+		0,
+	)
+
+	if err != nil {
+		return sessiontypes.Session{},
+			fmt.Errorf("GetSession: error getting the session for service %s app %s: %w",
+				serviceID, appAddr, err,
+			)
+	}
+
+	if session == nil {
+		return sessiontypes.Session{},
+			fmt.Errorf("GetSession: got nil session for service %s app %s: %w",
+				serviceID, appAddr, err,
+			)
+	}
+
+	return *session, nil
+}
+
+// ValidateRelayResponse validates the RelayResponse and verifies the supplier's signature.
+//
+// - Returns the RelayResponse, even if basic validation fails (may contain error reason).
+// - Verifies supplier's signature with the provided publicKeyFetcher.
+func (lfn *LazyFullNode) ValidateRelayResponse(
+	ctx context.Context,
+	supplierAddress sdk.SupplierAddress,
+	relayResponseBz []byte,
+) (*servicetypes.RelayResponse, error) {
+	relayResponse := &servicetypes.RelayResponse{}
+	if err := relayResponse.Unmarshal(relayResponseBz); err != nil {
+		return nil, err
+	}
+
+	if err := relayResponse.ValidateBasic(); err != nil {
+		// Even if the relay response is invalid, return it (may contain failure reason)
+		return relayResponse, err
+	}
+
+	supplierPubKey, err := lfn.GetAccountPubKey(
```

Comment:
Per my other comments, this business logic should be in `relay.go`.

Simply pass in the pubKey in there.

Reply:
It's been added as a method to the `GatewayClient`. Please LMK if you disagree with this change.